sdk/go/common/util/rpcutil/debug_interceptor.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Logs all gRPC converations in JSON format.
+//
+// To enable, PULUMI_DEBUG_GRPC must be set, pointing to a file to
+// append logs to.
+//
+// For example:
+//
+//     PULUMI_DEBUG_GPRC=grpc.json pulumi preview
+//     jq . grpc.json
+//
+func debugServerInterceptor() grpc.UnaryServerInterceptor {
+	return debugInterceptorInstance
+}
+
+type debugInterceptorLogEntry struct {
+	Method   string          `json:"method"`
+	Request  json.RawMessage `json:"request"`
+	Response json.RawMessage `json:"response"`
+	Errors   []string        `json:"errors,omitempty"`
+}
+
+type debugInterceptor struct {
+	logFile string
+	mutex   sync.Mutex
```

Comment:
Note to self: we might have multiple OS processes starting gRPC servers via Serve hitting this code (is this true? Or only pulumi process start servers and the rest are clients?). If that is the case, sync.Mutex is not enough, an to write to a single file there need to be file locks. I can throw one just in case. For the use cases I have in mind here terrible locking overhead is OK to get started with this.

Reply:
Pulumi is a Resource and Engine server, the language plugins are Language servers, providers are Provider servers, automation api programs are Language servers. So yes there are multiple processes starting servers.